samples/boards/nrf/s2ram/src/main.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2022 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <pm/pm.h>
+#include <pm/policy.h>
+
+#include <arch/common/pm_s2ram.h>
+
+#include <hal/nrf_vmc.h>
+#include <hal/nrf_power.h>
+#include <hal/nrf_regulators.h>
+#include <hal/nrf_gpio.h>
+
+#define RAM_N_BLOCK	(8)
+#define MASK_ALL_SECT	(VMC_RAM_POWER_S0RETENTION_Msk  | VMC_RAM_POWER_S1RETENTION_Msk  | \
+			 VMC_RAM_POWER_S2RETENTION_Msk  | VMC_RAM_POWER_S3RETENTION_Msk  | \
+			 VMC_RAM_POWER_S4RETENTION_Msk  | VMC_RAM_POWER_S5RETENTION_Msk  | \
+			 VMC_RAM_POWER_S6RETENTION_Msk  | VMC_RAM_POWER_S7RETENTION_Msk  | \
+			 VMC_RAM_POWER_S8RETENTION_Msk  | VMC_RAM_POWER_S9RETENTION_Msk  | \
+			 VMC_RAM_POWER_S10RETENTION_Msk | VMC_RAM_POWER_S11RETENTION_Msk | \
+			 VMC_RAM_POWER_S12RETENTION_Msk | VMC_RAM_POWER_S13RETENTION_Msk | \
+			 VMC_RAM_POWER_S14RETENTION_Msk | VMC_RAM_POWER_S15RETENTION_Msk)
+
+static void ram_retention_config(bool enable)
+{
+	uint32_t mask_sect = MASK_ALL_SECT;
+
+	for (size_t n = 0; n < RAM_N_BLOCK; n++) {
+		if (enable) {
+			nrf_vmc_ram_block_retention_set(NRF_VMC, n, mask_sect);
+		} else {
+			nrf_vmc_ram_block_retention_clear(NRF_VMC, n, mask_sect);
+		}
+	}
+}
```

Comment:
I think that the problem is that some user can selectively retain only a portion of the RAM, shutting down the remaining blocks. In this sample I'm retaining the whole SRAM for sake of simplicity but it really should be left to the user to decide what part of the SRAM to retain and what blocks to shut down.

We can discuss whether we want to forcefully retain the whole SRAM when `CONFIG_PM_S2RAM` is enabled though.

Reply:
hmm maybe this can be configured via DT? Btw, for s2ram to work correctly, do we need to always retain all blocks? If answer is yes, then `CONFIG_PM_S2RAM` should likely enable this code in `soc/`.

Reply:
Well, you really have to retain all the blocks that you are using. It is unlikely that you really need the **whole** SRAM, but being smart about that is not straightforward.

Reply:
moved to `soc/`